pallets/extrinsic-filter/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+pub enum OperationalMode {
+	/// when parachain runs normally
+	Normal,
+	/// when parachain traps in contigency situation
+	Safe,
+	/// when parachain is used for testing purpose
+	Test,
+}
+
+impl Default for OperationalMode {
+	fn default() -> Self {
+		OperationalMode::Normal
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The priviledged origin to perform all operations
+		type UpdateOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Filters that will be specified in runtime binding
+		type NormalModeFilter: Contains<Self::Call>;
+		type SafeModeFilter: Contains<Self::Call>;
+		type TestModeFilter: Contains<Self::Call>;
+
+		/// Weights
+		type WeightInfo: WeightInfo;
+	}
+
+	/// current mode, ValueQuery as it can't be None
+	#[pallet::storage]
+	#[pallet::getter(fn mode)]
+	pub type Mode<T: Config> = StorageValue<_, OperationalMode, ValueQuery>;
+
+	/// a tuple (pallet_name_bytes, Option<function_name_bytes>) to represent blocked extrinsics
+	/// if `Option<function_name_bytes>` is None, then all extrinsics in `pallet_name_bytes` are
+	/// blocked
+	#[pallet::storage]
+	#[pallet::getter(fn blocked_extrinsics)]
+	pub type BlockedExtrinsics<T: Config> =
+		StorageMap<_, Twox64Concat, (Vec<u8>, Vec<u8>), (), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// a new mode was just sent
+		ModeSet { new_mode: OperationalMode },
+		/// some extrinsics are blocked
+		ExtrinsicsBlocked { pallet_name_bytes: Vec<u8>, function_name_bytes: Option<Vec<u8>> },
+		/// some extrinsics are unblocked
+		ExtrinsicsUnblocked { pallet_name_bytes: Vec<u8>, function_name_bytes: Option<Vec<u8>> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Error when a given extrinsic cannot be blocked (e.g. this pallet)
+		CannotBlock,
+		/// Error during conversion bytes to utf8 string
+		CannotConvertToString,
+		/// Error when trying to block extrinsic more than once
+		ExtrinsicAlreadyBlocked,
+		/// Error when trying to unblock a non-existent extrinsic
+		ExtrinsicNotBlocked,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set the mode
+		/// Weights should be 2 DB writes: 1 for mode and 1 for event
+		#[pallet::weight(2 * T::DbWeight::get().write)]
+		pub fn set_mode(origin: OriginFor<T>, mode: OperationalMode) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			// we don't care which mode we are currently in,
+			// just set to the desired mode
+			Mode::<T>::put(mode);
```

Comment:
If the mode changed, should we clear all BlockedExtrinsics?

Reply:
> If the mode changed, should we clear all BlockedExtrinsics?

Good question. We could have implemented it in either way.
Currently this pallet is designed more for security reasons, so it's very conservative: we'd rather have extra (unchanged) blocked extrinsics than having unexpected whitelisted extrinsics.

However I'll write a comment explicitly about this and make a test case. 
Thank you.